From d0054adef2bafd38a0a5d375c91176b411f71ee8 Mon Sep 17 00:00:00 2001
From: Jing Liu <jing2.liu@intel.com>
Date: Wed, 5 Jan 2022 04:35:15 -0800
Subject: [PATCH 0828/1646] [Backport] kvm: x86: Exclude unpermitted xfeatures
 at KVM_GET_SUPPORTED_CPUID

mainline inclusion
from mainline-v5.17-rc1
commit 445ecdf79be0c71ca248f7611aeefceaea3ec59f
category: feature
bugzilla: https://gitee.com/openeuler/intel-kernel/issues/I5RQLJ

Reference: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=445ecdf79be0c71ca248f7611aeefceaea3ec59f

--------------------------------

KVM_GET_SUPPORTED_CPUID should not include any dynamic xstates in
CPUID[0xD] if they have not been requested with prctl. Otherwise
a process which directly passes KVM_GET_SUPPORTED_CPUID to
KVM_SET_CPUID2 would now fail even if it doesn't intend to use a
dynamically enabled feature. Userspace must know that prctl is
required and allocate >4K xstate buffer before setting any dynamic
bit.

Suggested-by: Paolo Bonzini <pbonzini@redhat.com>
Signed-off-by: Jing Liu <jing2.liu@intel.com>
Signed-off-by: Yang Zhong <yang.zhong@intel.com>
Message-Id: <20220105123532.12586-5-yang.zhong@intel.com>
Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>
Signed-off-by: Lin Wang <lin.x.wang@intel.com>
Signed-off-by: Jialin Zhang <zhangjialin11@huawei.com>

[Backport] x86/fpu: Fix inline prefix warnings

mainline inclusion
from mainline-v5.17-rc1
commit c862dcd199759d4a45e65dab47b03e3e8a144e3a
category: feature
bugzilla: https://gitee.com/openeuler/intel-kernel/issues/I5RQLJ

Reference: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=c862dcd199759d4a45e65dab47b03e3e8a144e3a

--------------------------------

Fix sparse warnings in xstate and remove inline prefix.

Fixes: 980fe2fddcff ("x86/fpu: Extend fpu_xstate_prctl() with guest permissions")
Signed-off-by: Yang Zhong <yang.zhong@intel.com>
Reported-by: kernel test robot <lkp@intel.com>
Message-Id: <20220113180825.322333-1-yang.zhong@intel.com>
Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>
Signed-off-by: Lin Wang <lin.x.wang@intel.com>
Signed-off-by: Jialin Zhang <zhangjialin11@huawei.com>
---
 Documentation/virt/kvm/api.rst | 4 ++++
 arch/x86/include/asm/fpu/api.h | 2 +-
 arch/x86/kernel/fpu/xstate.c   | 2 +-
 arch/x86/kvm/cpuid.c           | 9 ++++++---
 4 files changed, 12 insertions(+), 5 deletions(-)

diff --git a/Documentation/virt/kvm/api.rst b/Documentation/virt/kvm/api.rst
index f8eac13c935f..6a7d152aadcf 100644
--- a/Documentation/virt/kvm/api.rst
+++ b/Documentation/virt/kvm/api.rst
@@ -1632,6 +1632,10 @@ userspace capabilities, and with user requirements (for example, the
 user may wish to constrain cpuid to emulate older hardware, or for
 feature consistency across a cluster).
 
+Dynamically-enabled feature bits need to be requested with
+``arch_prctl()`` before calling this ioctl. Feature bits that have not
+been requested are excluded from the result.
+
 Note that certain capabilities, such as KVM_CAP_X86_DISABLE_EXITS, may
 expose cpuid features (e.g. MONITOR) which are not supported by kvm in
 its default configuration. If userspace enables such capabilities, it
diff --git a/arch/x86/include/asm/fpu/api.h b/arch/x86/include/asm/fpu/api.h
index b915167b86da..c0e83c871e7f 100644
--- a/arch/x86/include/asm/fpu/api.h
+++ b/arch/x86/include/asm/fpu/api.h
@@ -111,7 +111,7 @@ static inline void fpstate_free(struct fpu *fpu) { }
 /* fpstate-related functions which are exported to KVM */
 extern void fpstate_clear_xstate_component(struct fpstate *fps, unsigned int xfeature);
 
-extern inline u64 xstate_get_guest_group_perm(void);
+extern u64 xstate_get_guest_group_perm(void);
 
 /* KVM specific functions */
 extern bool fpu_alloc_guest_fpstate(struct fpu_guest *gfpu);
diff --git a/arch/x86/kernel/fpu/xstate.c b/arch/x86/kernel/fpu/xstate.c
index 70290089c506..3b447385ee14 100644
--- a/arch/x86/kernel/fpu/xstate.c
+++ b/arch/x86/kernel/fpu/xstate.c
@@ -1728,7 +1728,7 @@ static inline int xstate_request_perm(unsigned long idx, bool guest)
 }
 #endif  /* !CONFIG_X86_64 */
 
-inline u64 xstate_get_guest_group_perm(void)
+u64 xstate_get_guest_group_perm(void)
 {
 	return xstate_get_group_perm(true);
 }
diff --git a/arch/x86/kvm/cpuid.c b/arch/x86/kvm/cpuid.c
index 0e1cb0c584d2..87cd8e21967f 100644
--- a/arch/x86/kvm/cpuid.c
+++ b/arch/x86/kvm/cpuid.c
@@ -760,11 +760,13 @@ static inline int __do_cpuid_func(struct kvm_cpuid_array *array, u32 function)
 		 */
 		entry->eax = entry->ebx = entry->ecx = 0;
 		break;
-	case 0xd:
-		entry->eax &= supported_xcr0;
+	case 0xd: {
+		u64 guest_perm = xstate_get_guest_group_perm();
+
+		entry->eax &= supported_xcr0 & guest_perm;
 		entry->ebx = xstate_required_size(supported_xcr0, false);
 		entry->ecx = entry->ebx;
-		entry->edx &= supported_xcr0 >> 32;
+		entry->edx &= (supported_xcr0 & guest_perm) >> 32;
 		if (!supported_xcr0)
 			break;
 
@@ -811,6 +813,7 @@ static inline int __do_cpuid_func(struct kvm_cpuid_array *array, u32 function)
 			entry->edx = 0;
 		}
 		break;
+	}
 	case 0x12:
 		/* Intel SGX */
 		if (!kvm_cpu_cap_has(X86_FEATURE_SGX)) {
-- 
2.17.1

